common/utilities/time/waiting-on.h

```diff
@@ -0,0 +1,134 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2021 Intel Corporation. All Rights Reserved.
+
+#pragma once
+
+#include <mutex>
+#include <condition_variable>
+#include <thread>
+#include <atomic>
+
+
+namespace utilities {
+namespace time {
+
+// Helper class -- encapsulate a variable of type T that we want to wait on: another thread will set
+// it and signal when we can continue...
+//
+template< class T >
+class waiting_on
+{
+    // We need to be careful with timeouts: if we time out then the local waiting_on can go out of
+    // scope and then the thread cannot set anything or signal anyone! We get around this by using a
+    // shared_ptr & weak_ptr to manage access:
+    //
+public:
+    class wait_state_t
+    {
+        T _value;
+        std::condition_variable _cv;
+
+        friend class waiting_on;
+
+    public:
+        wait_state_t() = default;   // allow default ctor
+        wait_state_t( T const & t )
+            : _value( t )
+        {
+        }
+
+        operator T &() { return _value; }
+        operator T const &() const { return _value; }
+
+        T* operator->() { return &_value; }
+        T const* operator->() const { return &_value; }
+
+        // Set a new value and signal
+        void signal( T const & t )
+        {
+            _value = t;
+            signal();
+        }
+        // Signal with the current value
+        void signal()
+        {
+            _cv.notify_one();
+        }
+    };
+private:
+    std::shared_ptr< wait_state_t > _ptr;
+
+    // When we declare the signalling lambda for the other thread, we need to pass it the weak_ptr.
+    // This class wraps it up nicely, so you can write:
+    //     waiting_on< bool > invoked( false )
+    //     auto thread_function = [invoked = invoked.in_thread()]() {
+    //         invoked.signal( true );
+    //     }
+    //
+public:
+    class in_thread_
+    {
+        std::weak_ptr< class wait_state_t > const _ptr;
+
+    public:
+        in_thread_( waiting_on const& local )
+            : _ptr( local._ptr )
+        {
+        }
+
+        std::shared_ptr< wait_state_t > still_alive() const { return _ptr.lock(); }
+
+        // Wake up the local function (which is using wait_until(), presumable) with a new
+        // T value
+        void signal( T const& t ) const
+        {
+            if( auto wait_state = still_alive() )
+                wait_state->signal( t );

```

Comment:
The operation needs to be atomic to avoid race cond.

Reply:
The whole class is meant to be used in a master-slave kind of relationship: one thread running one function (and not multiple threads running one function). You're waiting on something to happen with a timeout. That's all.

The use-case of multiple threads all trying to signal the same waiting object is still possible -- this class does not decide for the user what the synchronization mechanism is. The user writes the lambda, which can worry about synchronization if it needs it. I don't think it's the job of the class.

I'll add the above to the class comment.